beacon-chain/sync/validate_aggregate_proof_test.go

```diff
@@ -0,0 +1,259 @@
+	data := &ethpb.AttestationData{}
+
+	wanted := "validator is not an aggregator for slot"
+	if err := validateSelection(ctx, beaconState, data, 0, sig.Marshal()); !strings.Contains(err.Error(), wanted) {
+		t.Error("Did not receive wanted error")
+	}
+}
+
+func TestVerifySelection_BadSignature(t *testing.T) {
+	ctx := context.Background()
+	deposits, _, privKeys := testutil.SetupInitialDeposits(t, 256)
+	beaconState, err := state.GenesisBeaconState(deposits, uint64(0), &ethpb.Eth1Data{BlockHash: make([]byte, 32)})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sig := privKeys[0].Sign([]byte{}, 0)
+	data := &ethpb.AttestationData{}
+
+	wanted := "could not validate slot signature"
+	if err := validateSelection(ctx, beaconState, data, 0, sig.Marshal()); !strings.Contains(err.Error(), wanted) {
+		t.Error("Did not receive wanted error")
+	}
+}
+
+func TestVerifySelection_CanVerify(t *testing.T) {
+	ctx := context.Background()
+	deposits, _, privKeys := testutil.SetupInitialDeposits(t, 256)
+	beaconState, err := state.GenesisBeaconState(deposits, uint64(0), &ethpb.Eth1Data{BlockHash: make([]byte, 32)})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	data := &ethpb.AttestationData{}
+	slotRoot, err := ssz.HashTreeRoot(data.Slot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	domain := helpers.Domain(beaconState.Fork, 0, params.BeaconConfig().DomainBeaconAttester)
+	sig := privKeys[0].Sign(slotRoot[:], domain)
+
+	if err := validateSelection(ctx, beaconState, data, 0, sig.Marshal()); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestValidateAggregateAndProof_NoBlock(t *testing.T) {
+	db := dbtest.SetupDB(t)
+	defer dbtest.TeardownDB(t, db)
+
+	att := &ethpb.Attestation{
+		Data: &ethpb.AttestationData{
+			Source: &ethpb.Checkpoint{Epoch: 0, Root: []byte("hello-world")},
+			Target: &ethpb.Checkpoint{Epoch: 0, Root: []byte("hello-world")},
+		},
+	}
+
+	aggregateAndProof := &pb.AggregateAndProof{
+		SelectionProof:  []byte{'A'},
+		Aggregate:       att,
+		AggregatorIndex: 0,
+	}
+
+	r := &RegularSync{
+		db:          db,
+		initialSync: &mockSync.Sync{IsSyncing: false},
+	}
+
+	wanted := "attestation points to a block which is not in the database"
+	if _, err := r.validateAggregateAndProof(context.Background(), aggregateAndProof, &p2ptest.MockBroadcaster{}, false); !strings.Contains(err.Error(), wanted) {
+		t.Error("Did not receive wanted error")
+	}
+}
+
+func TestValidateAggregateAndProof_NotWithinSlotRange(t *testing.T) {
+	db := dbtest.SetupDB(t)
+	defer dbtest.TeardownDB(t, db)
+
+	deposits, _, _ := testutil.SetupInitialDeposits(t, 256)
+	beaconState, err := state.GenesisBeaconState(deposits, uint64(0), &ethpb.Eth1Data{BlockHash: make([]byte, 32)})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	b := &ethpb.BeaconBlock{}
+	db.SaveBlock(context.Background(), b)
+	root, _ := ssz.SigningRoot(b)
+
+	aggBits := bitfield.NewBitlist(3)
+	aggBits.SetBitAt(0, true)
+	att := &ethpb.Attestation{
+		Data: &ethpb.AttestationData{
+			Slot:            1,
```

Comment:
Last request (I promise!) could you please add a slot for the future case? I identified a potential attack vector and it would be nice to have a unit test to protect against it. 

Perhaps a test with attestation data slot at max uint64 or some other large number.

Reply:
Np. Added a case for slot = ` 1<<64 - 1`